### PR TITLE
vkd3d: Implement DiscardResource.

### PR DESCRIPTION
**Pull request by doitsujin**
Tries to make use of our existing render pass transitions where possible, and otherwise just emits a barrier transitioning away from ``UNDEFINED``.

### CODE REVIEW

File: libs/vkd3d/resource.c [hunk 1]
```diff
@@ -1960,6 +1960,31 @@ struct d3d12_resource *unsafe_impl_from_ID3D12Resource(ID3D12Resource *iface)
...
+    const struct vkd3d_format *format = vkd3d_format_from_d3d12_resource_desc(resource->device, &resource->desc, 0);
+    VkImageSubresource subresource;
+
+    subresource.aspectMask = format->vk_aspect_mask;
+    subresource.mipLevel = subresource_idx % resource->desc.MipLevels;
+    subresource.arrayLayer = (subresource_idx / resource->desc.MipLevels) % resource->desc.DepthOrArraySize;
+
+    if (!all_aspects)
+    {
+        uint32_t i, plane_idx = subresource_idx / (resource->desc.MipLevels * resource->desc.DepthOrArraySize);
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:1974]: This seems a little sketchy for 3D images. Can they enter this path somehow?

File: libs/vkd3d/resource.c [hunk 2]
```diff
@@ -1960,6 +1960,31 @@ struct d3d12_resource *unsafe_impl_from_ID3D12Resource(ID3D12Resource *iface)
...
+
+        /* Clear the lower 'plane_idx' bits, then isolate the lowest bit.
+         * This works because aspect numbering between D3D12 and Vulkan
+         * is the same. */
+        for (i = 0; i < plane_idx; i++)
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:1979]: It is a bit hard to understand (without knowing the full implementation) why this is done.

File: libs/vkd3d/command.c [hunk 3]
```diff
@@ -1857,39 +1857,46 @@ static VkImageAspectFlags vk_writable_aspects_from_image_layout(VkImageLayout vk
...
-        const struct d3d12_resource *resource, const struct vkd3d_view *view)
+static int d3d12_command_list_find_attachment_view(struct d3d12_command_list *list,
+        const struct d3d12_resource *resource, const VkImageSubresourceLayers *subresource)
 {
     unsigned int i;
 
     if (list->dsv.resource == resource)
     {
-        const struct vkd3d_view *dsv = list->dsv.view;
+        struct vkd3d_view *dsv = list->dsv.view;
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:1867]: Why drop the const?

File: libs/vkd3d/command.c [hunk 4]
```diff
@@ -1857,39 +1857,46 @@ static VkImageAspectFlags vk_writable_aspects_from_image_layout(VkImageLayout vk
...
+                dsv->info.texture.layer_idx == subresource->baseArrayLayer &&
+                dsv->info.texture.layer_count == subresource->layerCount)
             return D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT;
     }
     else
     {
         for (i = 0; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; i++)
         {
-            const struct vkd3d_view *rtv = list->rtvs[i].view;
+            struct vkd3d_view *rtv = list->rtvs[i].view;
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:1878]: Ditto about const.

File: libs/vkd3d/command.c [hunk 5]
```diff
@@ -6287,7 +6355,77 @@ static void STDMETHODCALLTYPE d3d12_command_list_ClearUnorderedAccessViewFloat(d
...
+        return;
+    }
+
+    /* We can't meaningfully discard sub-regions of an image. If rects
+     * are specified, all specified subresources must have the same
+     * dimensions, so just base this off the first subresource */
+    if (!(full_discard = !region->NumRects))
+    {
+        vk_subresource = d3d12_resource_get_vk_subresource(texture, region->FirstSubresource, false);
+        full_rect = d3d12_get_image_rect(texture, vk_subresource.mipLevel);
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:6396]: I guess the reason you're not checking all subresources is the d3d12 requirement that all subresources touched by Discard must have the same dimensions?

File: libs/vkd3d/command.c [hunk 6]
```diff
@@ -6287,7 +6355,77 @@ static void STDMETHODCALLTYPE d3d12_command_list_ClearUnorderedAccessViewFloat(d
...
+
+        for (i = 0; i < region->NumRects && !full_discard; i++)
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:6398]: This seems a bit fishy, I would expect to see full_discard be initialized to true here, and if there is any non-full discard, it nopes out.

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:6398]: The way I read the code now, it's considered a full discard if just one rect is full.

File: libs/vkd3d/command.c [hunk 7]
```diff
@@ -2111,6 +2111,66 @@ static void d3d12_command_list_clear_attachment_deferred(struct d3d12_command_li
...
+    const struct vkd3d_format *format;
+    VkImageMemoryBarrier barrier;
+    VkPipelineStageFlags stages;
+    VkAccessFlags access;
+    VkImageLayout layout;
+
+    if (resource->desc.Flags & D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET)
+    {
+        stages = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+        access = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:2127]: What about read bits?

File: libs/vkd3d/command.c [hunk 8]
```diff
@@ -2111,6 +2111,66 @@ static void d3d12_command_list_clear_attachment_deferred(struct d3d12_command_li
...
+
+    /* Transitioning only one aspect of a depth-stencil image is not allowed */
+    format = vkd3d_format_from_d3d12_resource_desc(list->device, &resource->desc, 0);
+
+    if (format->vk_aspect_mask != subresource->aspectMask)
+        return;
+
+    barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    barrier.pNext = NULL;
+    barrier.srcAccessMask = 0;
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:2150]: If we're in a suspended render pass, is it safe to use 0 here for srcAccessMask?

File: libs/vkd3d/command.c (hunk 5 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:6396]: Yes.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:6398]: > it's considered a full discard if just one rect is full.

Yes, that's the idea and also done this way for clears. Discarding sub-rects doesn't have any effect when the entire subresource is cleared.

File: libs/vkd3d/command.c (hunk 7 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2127]: Do we need those? Reads are undefined after discard.

File: libs/vkd3d/command.c (hunk 8 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:2150]: We're discarding old content, so we have no reason to flush it.

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2150]: Technically, memory needs to be available to be able to safely do a layout transition, even from UNDEFINED. I guess the possible hazard would be that this chain of events could get mangled:

- Write something (A)
- Discard
- Write something (B)

The problem scenario is if there are some pending writes in A, which are only resolved after B writes resolve.

File: libs/vkd3d/command.c (hunk 7 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2127]: I suppose we can drop the read.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:6398]: Oh, wait. All subrects apply to all subresources?

File: libs/vkd3d/command.c (hunk 8 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:2150]: This seems a bit pedantic though. I'll be okay doing 0 here if it's commented why it's 0.

File: libs/vkd3d/command.c (hunk 6 above)
Comment (doitsujin) [reply libs/vkd3d/command.c:6398]: Yes, that's why the API requires to have all subresources to have the same dimension if discard rects are used.

File: libs/vkd3d/command.c (hunk 3 above)
Comment (doitsujin) [reply libs/vkd3d/command.c]: Re-added; this was part of a temporary change and is no longer needed.

File: libs/vkd3d/command.c (hunk 4 above)
Comment (doitsujin) [reply libs/vkd3d/command.c]: Same.

File: libs/vkd3d/resource.c (hunk 2 above)
Comment (doitsujin) [reply libs/vkd3d/resource.c:1980]: Changed the comment.

File: libs/vkd3d/resource.c (hunk 1 above)
Comment (doitsujin) [reply libs/vkd3d/resource.c]: Yeah, fixed.

File: libs/vkd3d/resource.c (hunk 2 above)
Comment (HansKristian-Work) [reply libs/vkd3d/resource.c:1980]: Ok, I think I get it now. Clever bithack indeed.

**Pull request merged**